plugins/upower.c

```diff
@@ -0,0 +1,276 @@
+#define UPOWER_BASE_BATT_PATH     "/org/freedesktop/UPower/devices/battery_"
+#define UPOWER_TOUCH_BATT_PATH    UPOWER_BASE_BATT_PATH "battery"
+
+static guint modem_watch;
+static guint upower_battery_watch;
+static guint upower_daemon_watch;
+static DBusConnection *connection;
+static GHashTable *modem_hfp_watches;
+static GList *modems;
+static int last_battery_val;
+
+static void emulator_battery_cb(struct ofono_atom *atom, void *data)
+{
+	struct ofono_emulator *em = __ofono_atom_get_data(atom);
+	int val = GPOINTER_TO_INT(data);
+
+	DBG("%s: calling set_indicator: %d", __func__, val);
+	ofono_emulator_set_indicator(em, OFONO_EMULATOR_IND_BATTERY, (int) val);
+}
+
+static gboolean battery_props_changed(DBusConnection *conn, DBusMessage *msg,
+				void *user_data)
+
+{
+	const char *iface;
+	const char *property;
+	DBusMessageIter iter, dict;
+	GList *list;
+	double percentage_val;
+	int battery_val;
+
+	DBG("%s:", __func__);
+
+	dbus_message_iter_init(msg, &iter);
+
+	/* 1st param is an 'iface' - check string type */
+	if (dbus_message_iter_get_arg_type(&iter) != DBUS_TYPE_STRING) {
+		ofono_error("%s: iface != TYPE_STRING!", __func__);
+		return FALSE;
+	}
+
+	dbus_message_iter_get_basic(&iter, &iface);
+
+	/* Check for UPower.Device iface */
+	if (g_str_equal(iface, UPOWER_DEVICE_INTERFACE) != TRUE) {
+		ofono_error("%s: wrong iface: %s!", __func__, iface);
+		return FALSE;
+	}
+
+	/* advance iter... */
+	if (!dbus_message_iter_next(&iter)) {
+		ofono_error("%s: advance iter failed!", __func__);
+		return FALSE;
+	}
+
+	/* @nd param is Dict; validate wrapper is an array */
+	if (dbus_message_iter_get_arg_type(&iter) != DBUS_TYPE_ARRAY) {
+		ofono_error("%s: type != ARRAY!", __func__);
+		return FALSE;
+	}
+
+	dbus_message_iter_recurse(&iter, &dict);
+
+	while (dbus_message_iter_get_arg_type(&dict) == DBUS_TYPE_DICT_ENTRY) {
+		DBusMessageIter entry, val;
+		const char *key;
+
+		dbus_message_iter_recurse(&dict, &entry);
+
+		if (dbus_message_iter_get_arg_type(&entry) != DBUS_TYPE_STRING) {
+			ofono_info("%s: key type != STRING!", __func__);
+			return FALSE;
+		}
+
+		dbus_message_iter_get_basic(&entry, &key);
+
+		if (g_str_equal(key, "Percentage") != TRUE) {
+			ofono_info("%s: key != 'Percentage': %s", __func__, key);
+
+			dbus_message_iter_next(&dict);
+			continue;
+		}
+
+		ofono_info("%s: key == 'Percentage'", __func__);
+
+		dbus_message_iter_next(&entry);
+
+		if (dbus_message_iter_get_arg_type(&entry) != DBUS_TYPE_VARIANT) {
+			ofono_info("%s: 'Percentage' val != VARIANT", __func__);
+			return FALSE;
+		}
+
+		dbus_message_iter_recurse(&entry, &val);
+
+		if (dbus_message_iter_get_arg_type(&val) != DBUS_TYPE_DOUBLE) {
+			ofono_info("%s: 'Percentage' val != DOUBLE", __func__);
+			return FALSE;
+		}
+
+		dbus_message_iter_get_basic(&val, &percentage_val);
+		break;
+	}
+
+	if (percentage_val == 0.00) {
+		battery_val = 0;
+	} else if (percentage_val > 0.00 && percentage_val <= 100.00) {
```

Comment:
Why instead of battery_val==0 only when percentage is 0, and then define 5 bins of size 20 you don't define 6 equally sized bins (with battery_val among 0..5) ?